zeroize/build.rs

```diff
@@ -13,42 +15,113 @@ fn main() {
 
 /// Support for building the `explicit_bzero.c` backport.
 /// Only used when the `linux-backport` cargo feature is enabled and the
-/// installed glibc version is < 2.25.
+/// installed glibc version is < 2.25 or musl-libc version is < 1.1.20.
 #[cfg(all(feature = "linux-backport", target_os = "linux"))]
 mod linux {
     use super::cc;
+    use super::semver::Version;
     use std::process::Command;
 
     /// First version of glibc to support `explicit_bzero()`
-    const GLIBC_WITH_EXPLICIT_BZERO: &str = "2.25";
+    const GLIBC_WITH_EXPLICIT_BZERO: &str = "2.25.0";
 
-    /// Build `src/os/linux/explicit_bzero_backport.c` using the `cc` crate
-    pub fn build_explicit_bzero_backport() {
-        let glibc_version = get_glibc_version();
-
-        // Hax: this probably isn't the best use of floats
-        if glibc_version.parse::<f32>().unwrap() < GLIBC_WITH_EXPLICIT_BZERO.parse::<f32>().unwrap()
-        {
-            cc::Build::new()
-                .file("src/os/linux/explicit_bzero_backport.c")
-                .compile("explicit_bzero");
-        }
+    /// First version of musl-libc to support `explicit_bzero()`
+    const MUSL_WITH_EXPLICIT_BZERO: &str = "1.1.20";
+
+    enum StdLibrary {
+        /// GNU C standard library
+        GNU(Version),
+
+        /// Musl C standard library
+        Musl(Version),
+
+        /// Unsupported standard library
+        Unsupported,
     }
 
-    /// Get the current glibc version (vicariously by querying `/usr/bin/ldd`)
-    fn get_glibc_version() -> String {
-        let output = Command::new("/usr/bin/ldd")
-            .arg("--version")
-            .output()
-            .unwrap();
+    impl StdLibrary {
+        /// Build backports if necessary
+        fn should_build_explicit_bzero(&self) -> Option<bool> {
+            match self {
+                StdLibrary::GNU(ver) => Some(ver < &Version::parse(GLIBC_WITH_EXPLICIT_BZERO).unwrap()),
+                StdLibrary::Musl(ver) => Some(ver < &Version::parse(MUSL_WITH_EXPLICIT_BZERO).unwrap()),
+                StdLibrary::Unsupported => None,
+            }
+        }
+
+        /// Resolve the version of the installed C standard library
+        fn resolve() -> Self {
+            let output = Command::new("/usr/bin/ldd")
+                .arg("--version")
+                .output()
+                .unwrap();
```

Comment:
It looks like the `Command` invocations, along with converting stdout/stderr into a string, are duplicated in `get_glibc_version()` and `get_musl_version()`.

Perhaps extract that into something like `struct LddVersion { stdout: String, stderr: String, success: true }` with a `::get() -> LddVersion` or thereabout method that takes care of running the command and converting stdout/stderr to strings?

Reply:
@tarcieri I think it looks good now.